Sources/NetworkProtection/Recovery/FailureRecoveryHandler.swift

```diff
@@ -0,0 +1,198 @@
+                    eventHandler(.completed(.healthy))
+                default:
+                    eventHandler(.failed(error))
+                    throw error
+                }
+            }
+        }
+    }
+
+    func stop() {
+        task = nil
+    }
+
+    private func makeRecoveryAttempt(
+        to lastConnectedServer: NetworkProtectionServer,
+        includedRoutes: [IPAddressRange],
+        excludedRoutes: [IPAddressRange],
+        isKillSwitchEnabled: Bool
+    ) async throws -> NetworkProtectionDeviceManagement.GenerateTunnelConfigResult {
+        let serverSelectionMethod: NetworkProtectionServerSelectionMethod = .failureRecovery(serverName: lastConnectedServer.serverName)
+        let configurationResult: NetworkProtectionDeviceManagement.GenerateTunnelConfigResult
+
+        configurationResult = try await deviceManager.generateTunnelConfiguration(
+            selectionMethod: serverSelectionMethod,
+            includedRoutes: includedRoutes,
+            excludedRoutes: excludedRoutes,
+            isKillSwitchEnabled: isKillSwitchEnabled,
+            regenerateKey: false
+        )
+        os_log("🟢 Failure recovery fetched new config.", log: .networkProtectionServerFailureRecoveryLog, type: .info)
+
+        let newServer = configurationResult.server
+
+        os_log(
+            "🟢 Failure recovery - originalServerName: %{public}s, newServerName: %{public}s, originalAllowedIPs: %{public}s, newAllowedIPs: %{public}s",
+            log: .networkProtection,
+            type: .info,
+            lastConnectedServer.serverName,
+            newServer.serverName,
+            String(describing: lastConnectedServer.allowedIPs),
+            String(describing: newServer.allowedIPs)
+        )
+
+        guard lastConnectedServer.shouldReplace(with: newServer) else {
+            os_log("🟢 Server failure recovery not necessary.", log: .networkProtectionServerFailureRecoveryLog, type: .info)
+            throw FailureRecoveryError.noRecoveryNecessary
+        }
+
+        return configurationResult
+    }
+
+    private func incrementalPeriodicChecks(
+        _ config: RetryConfig,
+        action: @escaping () async throws -> Void
+    ) async {
+        let task = Task(priority: .background) {
+            var currentDelay = config.initialDelay
+            var count = 0
+            repeat {
+                do {
+                    try await action()
+                    os_log("🟢 Failure recovery success!", log: .networkProtectionServerFailureRecoveryLog, type: .info)
+                    return
+                } catch {
+                    os_log("🟢 Failure recovery failed. Retrying...", log: .networkProtectionServerFailureRecoveryLog, type: .info)
+                    try? await Task.sleep(interval: currentDelay)
+                }
+                count += 1
+                currentDelay = min((currentDelay * config.factor), config.maxDelay)
+            } while count < config.times
```

Comment:
What's the expected behavior if this is exceeded?

My understanding from what I'm seeing is that:
1. We try recovery 5 times (Each time we duplicate the previous delay, but never wait more than 5 minutes)
2. Once the attempts all fail, we bail out, and the failure monitor will keep checking periodically, but we won't do any more active recovery.

Is this correct?

Reply:
Yeah, this is correct. I was initially throwing an error that gets bubbled up to the PacketTunnelProvider. But I wasn’t reacting to it in any way so decided it wasn’t necessary. I guess we could send another pixel in that case, but I’d rather not just now as it wasn’t in the original scope and will require another privacy triage etc.